app/src/main/java/com/github/se/cyrcle/model/authentication/AuthenticationRepositoryGoogle.kt

```diff
@@ -0,0 +1,105 @@
+import kotlinx.coroutines.tasks.await
+
+/**
+ * Implementation of the Authenticator interface This class uses the Firebase Authentication SDK to
+ * authenticate users
+ */
+@ActivityScoped
+class AuthenticationRepositoryGoogle
+@Inject
+constructor(
+    @ActivityContext private val context: Context,
+    private val auth: FirebaseAuth,
+) : AuthenticationRepository {
+
+  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  var credentialManager = CredentialManager.create(context)
+
+  /**
+   * Signs in a User
+   *
+   * @param onSuccess a function to be called once the user is authenticated
+   * @param onFailure a function to be called if the authentication fails
+   */
+  override fun authenticate(onSuccess: (String) -> Unit, onFailure: (Exception) -> Unit) {
+    val webClientId = context.getString(R.string.default_web_client_id)
+
+    // Generate a nonce for security (a random number used once)
+    val ranNonce: String = UUID.randomUUID().toString()
+    val bytes: ByteArray = ranNonce.toByteArray()
+    val md: MessageDigest = MessageDigest.getInstance("SHA-256")
+    val digest: ByteArray = md.digest(bytes)
+    val hashedNonce: String = digest.fold("") { str, it -> str + "%02x".format(it) }
+
+    // Setup the options for the credential request
+    val signInWithGoogleOption =
+        GetSignInWithGoogleOption.Builder(webClientId).setNonce(hashedNonce).build()
+
+    val request = GetCredentialRequest.Builder().addCredentialOption(signInWithGoogleOption).build()
+
+    CoroutineScope(Dispatchers.Main).launch {
+      try {
+        val credential = credentialManager.getCredential(context, request).credential
+
+        if (credential is CustomCredential &&
+            credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
+
+          // Use googleIdTokenCredential and extract id to validate the user
+          val googleIdTokenCredential = GoogleIdTokenCredential.createFrom(credential.data)
+          val authCredential =
+              GoogleAuthProvider.getCredential(googleIdTokenCredential.idToken, null)
+          val authResult = auth.signInWithCredential(authCredential).await()
+
+          onSuccess(authResult.user!!.uid)
+        } else {
+          onFailure(Exception("Unexpected type of credential"))
+        }
+      } catch (e: Exception) {
+        onFailure(e)
+      }
+    }
+  }
+
+  /**
+   * Sign in a user anonymously
+   *
+   * @param onComplete a function to be called once the user is loged in anonymously
+   */
+  override fun authenticateAnonymously(onComplete: () -> Unit) {
+    runBlocking { auth.signInAnonymously().await() }
+    onComplete()
+  }
+
+  /**
+   * Sign out the user
+   *
+   * @param onComplete a function to be called once the user logs out
+   */
+  override fun signOut(onComplete: () -> Unit) {
+    runBlocking { credentialManager.clearCredentialState(ClearCredentialStateRequest()) }
+    auth.signOut()
+    onComplete()
+  }
```

Comment:
For `authenticateAnonymously` and `signOut` functions, I'm not sure to understand why they need an `onComplete` , as it is simply called at the end of both function (there is no "clean-up" at the end).

Reply:
This is to keep linearity during execution, since the repository methods are asynchronous, I think it could be important to update the UI only once the signIn/Out are made in the backend.

Reply:
Oh, I didn't thought of this